module-services/service-cellular/Pdu.cpp

```diff
@@ -0,0 +1,972 @@
+        MmsMReadOrigInd,     // m-read-orig-ind
+        MmsMForwardReq,      // m-forward-req
+        MmsMForwardComf,     // m-forward-conf
+        MmsMMboxStoreReq,    // m-mbox-store-req
+        MmsMMboxStoreConf,   // m-mbox-store-conf
+        MmsMMboxViewReq,     // m-mbox-view-req
+        MmsMMboxViewConf,    // m-mbox-view-conf
+        MmsMMboxUploadReq,   // m-mbox-upload-req
+        MmsMMboxUploadConf,  // m-mbox-upload-conf
+        MmsMMboxDeleteReq,   // m-mbox-delete-req
+        MmsMMboxDeleteConf,  // m-mbox-delete-conf
+        MmsMMboxDescr,       // m-mbox-descr
+        MmsMDeleteReq,       // m-delete-req
+        MmsMDeleteConf,      // m-delete-conf
+        MmsMCancelReq,       // m-cancel-req
+        MmsMCancelConf       // m-cancel-conf
+    };
+
+} // namespace pdu::constants
+
+namespace pdu
+{
+
+    // A range of characters
+    struct CharRange
+    {
+        using Iterator = std::string::const_iterator;
+
+        CharRange() : m_begin(), m_end()
+        {}
+
+        CharRange(Iterator begin, Iterator end) : m_begin(begin), m_end(end)
+        {
+            assert(m_begin <= m_end);
+        }
+
+        Iterator begin() const
+        {
+            return m_begin;
+        }
+        Iterator end() const
+        {
+            return m_end;
+        }
+        bool empty() const
+        {
+            return m_begin == m_end;
+        }
+        std::uint32_t size() const
+        {
+            return m_end - m_begin;
+        }
+
+        std::string str() const
+        {
+            return std::string(m_begin, m_end);
+        }
+
+        friend bool operator==(CharRange const &rng, std::string const &str)
+        {
+            return rng.size() == str.size() && std::equal(rng.m_begin, rng.m_end, str.begin());
+        }
+
+        friend bool operator==(std::string const &str, CharRange const &rng)
+        {
+            return rng == str;
+        }
+
+        friend bool operator==(CharRange const &rng, const char *cstr)
+        {
+            return cstr[rng.size()] == '\0' && std::equal(rng.m_begin, rng.m_end, cstr);
+        }
+
+        friend bool operator==(const char *cstr, CharRange const &rng)
+        {
+            return rng == cstr;
+        }
+
+      protected:
+        Iterator m_begin;
+        Iterator m_end;
+    };
+
+    // A buffer representing input octets
+    struct Octets : CharRange
+    {
+        Octets() = default;
+
+        Octets(Iterator begin, Iterator end) : CharRange(begin, end)
+        {}
+
+        bool peek(std::uint8_t &result) const
+        {
+            if (m_begin != m_end) {
+                result = *m_begin;
+                return true;
+            }
+            return false;
+        }
+
+        bool next(std::uint8_t &result)
+        {
+            if (m_begin != m_end) {
+                result = *m_begin++;
+                return true;
+            }
+            return false;
+        }
+
+        void eat()
```

Comment:
just wondering - shouldn't eat/consume method somehow take out the data? as it's being eaten/consumed. Here you just move the pointer further, without altering the data

Reply:
Yes, `next()` advances the iterator and returns the data. `eat()` is there to only advance the iterator after `peek()`. There is also `eatN()` used to advance the iterator by N bytes if e.g. a header is detected that is not supported and can be ignored. If you think some other naming would be preferable I'm open to suggestion.

Reply:
I'd say that `proceed` or something like this would be more appropriate here - when I see `eat` function I'd assume, that the data is somehow lost

Reply:
What do you think about `ignore()` which is a name used in `istream`? I could also rename `next()` with `get()` to be consistent.

Reply:
as discussed on private channel - ignore and next will be fine ;)